src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttribute.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class TagHelperAttribute<TValue>
+    {
+        private string _key;
+
+        public TagHelperAttribute()
+        {
+        }
+
+        public TagHelperAttribute([NotNull] string key)
+        {
+            Key = key;
+        }
+
+        public TagHelperAttribute([NotNull] string key, TValue value)
+        {
+            Key = key;
+            Value = value;
+        }
+
+        public string Key
+        {
+            get
+            {
+                return _key;
+            }
+            set
+            {
+                // Keys aren't allowed to be null.
+                if (value == null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
```

Comment:
fyi (since this is still [Design]) [guidelines](https://github.com/aspnet/Home/wiki/Engineering-guidelines#argument-null-checking) say to use `[param: NotNull]` above the setter instead.